go-chaos/cmd/backup.go

```diff
@@ -58,12 +71,121 @@ var waitForBackupCommand = &cobra.Command{
 	RunE:  waitForBackup,
 }
 
-func takeBackup(cmd *cobra.Command, args []string) error {
+var restoreBackupCommand = &cobra.Command{
+	Use:   "restore",
+	Short: "Restore from a given backup id",
+	RunE:  restoreFromBackup,
+}
+
+func setupBackup(cmd *cobra.Command, _ []string) error {
+	k8Client, err := internal.CreateK8Client()
+	if err != nil {
+		panic(err)
+	}
+
+	namespace := k8Client.GetCurrentNamespace()
+
+	err = k8Client.PauseReconciliation()
+	if err != nil {
+		return err
+	}
+
+	_, err = createBackupSecret(cmd, k8Client, namespace)
+	if err != nil {
+		return err
+	}
+
+	err = setupStatefulSetForBackups(cmd, err, k8Client, namespace)
+	if err != nil {
+		return err
+	}
+	err = setupGatewayForBackups(cmd, err, k8Client, namespace)
+	return err
+}
+
+func setupStatefulSetForBackups(cmd *cobra.Command, err error, k8Client internal.K8Client, namespace string) error {
+	sfs, err := k8Client.GetZeebeStatefulSet()
+	if err != nil {
+		return err
+	}
+
+	sfsEnvFrom := sfs.Spec.Template.Spec.Containers[0].EnvFrom
+	sfs.Spec.Template.Spec.Containers[0].EnvFrom = append(sfsEnvFrom, core.EnvFromSource{SecretRef: &core.SecretEnvSource{LocalObjectReference: core.LocalObjectReference{Name: "zeebe-backup-store-s3"}}})
+
+	sfsEnv := sfs.Spec.Template.Spec.Containers[0].Env
+	sfs.Spec.Template.Spec.Containers[0].Env = append(
+		sfsEnv,
+		core.EnvVar{Name: "ZEEBE_BROKER_DATA_BACKUP_STORE", Value: "S3"},
+		core.EnvVar{Name: "ZEEBE_BROKER_EXPERIMENTAL_FEATURES_ENABLEBACKUP", Value: "true"},
+		core.EnvVar{Name: "MANAGEMENT_ENDPOINTS_WEB_EXPOSURE_INCLUDE", Value: "*"},
+		core.EnvVar{Name: "MANAGEMENT_ENDPOINTS_BACKUPS_ENABLED", Value: "true"},
+	)
+
+	_, err = k8Client.Clientset.AppsV1().StatefulSets(namespace).Update(cmd.Context(), sfs, meta.UpdateOptions{})
+	return err
+}
+
+func setupGatewayForBackups(cmd *cobra.Command, err error, k8Client internal.K8Client, namespace string) error {
+	saasGatewayLabels := meta.LabelSelector{
+		MatchLabels: map[string]string{"app.kubernetes.io/component": "standalone-gateway"},
+	}
+	var gatewayDeployments *apps.DeploymentList
+
+	gatewayDeployments, err = k8Client.Clientset.AppsV1().Deployments(namespace).List(cmd.Context(), meta.ListOptions{LabelSelector: labels.Set(saasGatewayLabels.MatchLabels).String()})
+	if err != nil {
+		return err
+	}
+	if len(gatewayDeployments.Items) == 0 {
+		selector := meta.LabelSelector{
+			MatchLabels: map[string]string{"app.kubernetes.io/component": "zeebe-gateway"},
+		}
+		gatewayDeployments, err = k8Client.Clientset.AppsV1().Deployments(namespace).List(
+			cmd.Context(),
+			meta.ListOptions{LabelSelector: labels.Set(selector.MatchLabels).String()},
+		)
+		if err != nil {
+			return err
+		}
+	}
+	gateway := gatewayDeployments.Items[0]
```

Comment:
Until here we have the same code already somewhere else (**checkIfGatewaysAreRunning**). But I can also clean that up later.